src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -0,0 +1,467 @@
+#include "irr/asset/CGraphicsPipelineLoaderMTL.h"
+
+using namespace irr;
+using namespace asset;
+
+bool CGraphicsPipelineLoaderMTL::isALoadableFileFormat(io::IReadFile* _file) const
+{
+    if (!_file)
+        return false;
+
+    const size_t prevPos = _file->getPos();
+
+    _file->seek(0ull);
+
+    std::string mtl;
+    mtl.resize(_file->getSize());
+    _file->read(mtl.data(), _file->getSize());
+    _file->seek(prevPos);
+
+    return mtl.find("newmtl") != std::string::npos;
+}
+
+namespace
+{
+    core::smart_refctd_ptr<ICPUPipelineLayout> makePipelineLayoutFromMtl(const CMTLPipelineMetadata::SMtl& _mtl)
+    {
+        const size_t textureCnt = std::count_if(_mtl.maps, _mtl.maps + CMTLPipelineMetadata::SMtl::EMP_REFL_POSX + 1u, [](const std::string& _path) { return !_path.empty(); });
+        auto bindings = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUDescriptorSetLayout::SBinding>>(textureCnt);
+
+        ICPUDescriptorSetLayout::SBinding bnd;
+        bnd.count = 1u;
+        bnd.stageFlags = ICPUSpecializedShader::ESS_FRAGMENT;
+        bnd.type = EDT_COMBINED_IMAGE_SAMPLER;
+        std::fill(bindings->begin(), bindings->end(), bnd);
+
+        ICPUSampler::SParams smplParams;
+        smplParams.TextureWrapU = smplParams.TextureWrapV = smplParams.TextureWrapW = ICPUSampler::ETC_REPEAT;
+        smplParams.BorderColor = ICPUSampler::ETBC_FLOAT_TRANSPARENT_BLACK;
+        smplParams.MinFilter = smplParams.MaxFilter = ICPUSampler::ETF_LINEAR;
+        smplParams.MipmapMode = ICPUSampler::ESMM_LINEAR;
+        smplParams.AnisotropicFilter = 0u;
+        smplParams.CompareEnable = 0u;
+        smplParams.LodBias = 0.f;
+        smplParams.MinLod = -1000.f;
+        smplParams.MaxLod = 1000.f;
+
+        core::smart_refctd_ptr<ICPUSampler> samplers[2];
+        samplers[0] = core::make_smart_refctd_ptr<ICPUSampler>(smplParams);
+        smplParams.TextureWrapU = smplParams.TextureWrapV = smplParams.TextureWrapW = ICPUSampler::ETC_CLAMP_TO_BORDER;
+        samplers[1] = core::make_smart_refctd_ptr<ICPUSampler>(smplParams);
+        for (uint32_t i = 0u, j = 0u; i <= CMTLPipelineMetadata::SMtl::EMP_REFL_POSX; ++i)
+        {
+            if (!_mtl.maps[i].empty())
+            {
+                (*bindings)[j].binding = j;
+
+                const uint32_t clamp = (_mtl.clamp >> i) & 1u;
+                (*bindings)[j].samplers = samplers + clamp;
+
+                ++j;
+            }
+        }
+
+        auto dsLayout = core::make_smart_refctd_ptr<ICPUDescriptorSetLayout>(bindings->begin(), bindings->end());
+        //ds with textures for material goes to set=3
```

Comment:
set 3 is per mesh, set 2 is per material IIRC

0 = global / per-frame
1 = per-view/camera
2 = material/shader
3 = object/mesh/draw

Reply:
2 was for "mesh scene node", 3 for meshbuffer/draw. CPU/GPUmeshbuffer holds DS that ought to be used as set=3

Reply:
I see the following pairing (not gospel, approximate)
<global,0>
<renderpass,1>
<shader/renderpassindependentpipeline,2>
<draw command,3>

so i guess if you're not switching shaders when you're switching textures, it can do in set 3.